### DISCUSSION TITLE
Remove `CS0556` compiler error in user-defined conversion in generic type with specified T

### DISCUSSION THREAD

**Discussion by Vixenka** [dotnet/csharplang#6268] [Language Ideas]
# Introduction
Together with C# 11 and .NET 7, [INumber\<T>](https://docs.microsoft.com/en-us/dotnet/api/system.numerics.inumber-1?view=net-7.0) and similar interfaces will appear, which allows for generic math.

This will allow us for generic structures of multidimensional vectors, which helps not to duplicate the code for writing similar types like [Vector2](https://docs.unity3d.com/ScriptReference/Vector2.html) and [Vector2Int](https://docs.unity3d.com/ScriptReference/Vector2Int.html) in Unity engine.

It would be nice to be able to cast generic vector to another type generic vector using user-defined conversion like:
```cs
Vector2<float> floatVector = ...;
Vector2<int> intVector = (Vector2<int>)floatVector;
```
Which could be defined this way:
```cs
public record struct Vector2<T>(T X, T Y) where T : INumber<T>
{
    public static explicit operator Vector2<int>(Vector2<float> vector)
    {
        return new Vector2<int>((int)vector.X, (int)vector.Y);
    }
}
```

# Problem
Currently this is impossible because this conversion definition results in a compiler error [CS0556](https://docs.microsoft.com/en-us/dotnet/csharp/misc/cs0556). Which does not really make sense, since the `Vector2<T>` type is the most enclosing type from programmer's perspective for types created automatically from a generic type.

# Alternatives
Currently this is possible to perform a similar conversion using extension methods, for example:
```cs
public static class Vector2Extensions
{
   public static Vector2<int> CastToInt(this Vector2<float> vector)
   {
        return new Vector2<int>((int)vector.X, (int)vector.Y);
   }
}
```
With the usage:
```cs
Vector2<float> floatVector = ...;
Vector2<int> intVector = floatVector.CastToInt();
```
which is less natural in my opinion.

# Solution
My suggestion for solving this problem is to remove compiler error [CS0556](https://docs.microsoft.com/en-us/dotnet/csharp/misc/cs0556) in this case.

# Implementation problems
Additionally skipping this should not require changes in the .NET runtime, only in the language compiler. Also, it does not cause problems mentioned in a [similar discussion](https://github.com/dotnet/csharplang/discussions/515), which introduces more liberal changes, where the compiler is not supposed to know which operator to choose.

More problematic is if we were to create a similar operator based on the generic type like here:
```cs
public record struct Vector2<T>(T X, T Y) where T : INumber<T>
{
    // Limited to float user-defined conversion to Vector2<int>.
    public static explicit operator Vector2<int>(Vector2<float> vector)
    {
        return new Vector2<int>((int)vector.X, (int)vector.Y);
    }

    // Generic user-defined conversion to Vector2<int>.
    public static explicit operator Vector2<int>(Vector2<T> vector)
    {
        throw new NotImplementedException();
    }
}
```
This causes the compiler to receive two identical operators.

To solve this problem, the compiler could use the operator that best matches the type it is working on. That is, it would omit or replace the content of the generated code based on the generic pattern with a more matching user-defined operator.

**Comment by svick** [comment]
> More problematic is if we were to create a similar operator based on the generic type

I think an analogous situation exists today: converting from e.g. `Vector<float>` or `Vector<T>` to `SomeType<T>`:

```c#
struct SomeType<T>
{
    public static explicit operator SomeType<T>(Vector<float> vector);
    public static explicit operator SomeType<T>(Vector<T> vector);
}
```

In this case, [the compiler produces an error saying that the overloads are ambiguous](https://sharplab.io/#v2:EYLgtghglgdgNAFxAJwK4wD4AEAMACLARgDoA5VMAU2SgGMBnAbgFgAoNgfTwF48AKAMoB7KgBUAngAdKAHgBmAGyEQEAPgCUMSgHc8ANUq0EQ5PKUrVfdS3at6CNEbzCxU2aNV5tAC2qU8onggePaOCGwA3mx4MQQAzASEAGx4lAAekgp0UAh4QtLIKibOIpQS0jIefAZGJmbKangAbobGyOrRsV3cngAmlHIQqAoINp0xWAlEKemZ2bn51EXIJa4VVTVtlZ4tte3jXTx9A0MjNgC+QA===).

**Reply by Vixenka** [reply]
Ok, so actually this implementation problem does not apply to my proposal. This only makes it easier to implement my proposal.
